--- ../gnu_src/binutils-2.41/cpu/or1korbis.cpu	2023-07-03 01:00:00.000000000 +0200
+++ ./or1korbis.cpu	2023-09-20 10:38:23.943355104 +0200
@@ -22,6 +22,7 @@
 ; Hardware for immediate operands
 (dnh h-simm16      "16-bit signed immediate"   ((MACH ORBIS-MACHS)) (immediate (INT 16)) () () ())
 (dnh h-uimm16      "16-bit unsigned immediate" ()                   (immediate (UINT 16)) () () ())
+(dnh h-uimm8       "8-bit unsigned immediate" ()                    (immediate (UINT 8)) () () ())
 (dnh h-uimm6       "6-bit unsigned immediate"  ()                   (immediate (UINT 6)) () () ())
 
 ; Hardware for the (internal) atomic registers
@@ -61,6 +62,7 @@
 (dnf f-resv-10-7   "resv-10-7"                 ((MACH ORBIS-MACHS) RESERVED) 10 7)
 (dnf f-resv-10-3   "resv-10-3"                 ((MACH ORBIS-MACHS) RESERVED) 10 3)
 (dnf f-resv-10-1   "resv-10-1"                 ((MACH ORBIS-MACHS) RESERVED) 10 1)
+(dnf f-resv-7-8    "resv-7-8"                  ((MACH ORBIS-MACHS) RESERVED) 7 8)
 (dnf f-resv-8-1    "resv-8-1"                  ((MACH ORBIS-MACHS) RESERVED) 8 1)
 (dnf f-resv-7-4    "resv-7-4"                  ((MACH ORBIS-MACHS) RESERVED) 7 4)
 (dnf f-resv-5-2    "resv-5-2"                  ((MACH ORBIS-MACHS) RESERVED) 5 2)
@@ -97,6 +99,7 @@
 ; Immediates.
 (dnf f-uimm16    "uimm16"                      ((MACH ORBIS-MACHS))          15 16)
 (df  f-simm16    "simm16"                      ((MACH ORBIS-MACHS) SIGN-OPT) 15 16 INT #f #f)
+(dnf f-uimm8     "uimm8"                       ((MACH ORBIS-MACHS))          7  8)
 (dnf f-uimm6     "uimm6"                       ((MACH ORBIS-MACHS))          5  6) ;; shroti
 
 (define-multi-ifield
@@ -318,6 +321,7 @@
 (dnop atomic-address    "atomic address"                   ((MACH ORBIS-MACHS) SEM-ONLY) h-atomic-address    f-nil)
 
 (dnop uimm6             "uimm6"                            ((MACH ORBIS-MACHS))          h-uimm6             f-uimm6)
+(dnop uimm8             "uimm8"                            ((MACH ORBIS-MACHS))          h-uimm8             f-uimm8)
 
 (dnop rD                "destination register"             ((MACH ORBIS-MACHS))          h-gpr               f-r1)
 (dnop rA                "source register A"                ((MACH ORBIS-MACHS))          h-gpr               f-r2)
@@ -557,6 +561,95 @@
      ()
 )
 
+; Nios custom instructions
+(dni l-nios_rrr "nios_rrr reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_rrr $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 0) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_rrc "nios_rrc reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_rrc $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 2) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_rcr "nios_rcr reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_rcr $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 4) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_rcc "nios_rcc reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_rcc $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 6) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_crr "nios_crr reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_crr $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 1) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_crc "nios_crc reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_crc $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 3) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_ccr "nios_ccr reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_ccr $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 5) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l-nios_ccc "nios_ccc reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.nios_ccc $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST1 rD rA rB (f-resv-10-3 7) uimm8)
+     (nop)
+     ()
+     )
+; special instruction for the or1400
+(dni l-jumps "jumps"
+     ((MACH ORBIS-MACHS))
+     "l.jumps"
+     (+ OPC_CUST2 (f-resv-25-26 #x30) )
+     (nop)
+     ()
+     )
+
+; cas and swap
+(dni l_cas "cas reg/reg/reg/uimm8"
+     ((MACH ORBIS-MACHS))
+     "l.cas $rD,$rA,$rB,$uimm8"
+     (+ OPC_CUST2 rD rA rB (f-resv-10-3 0) uimm8)
+     (nop)
+     ()
+     )
+
+(dni l_swap "swap reg/reg/reg"
+     ((MACH ORBIS-MACHS))
+     "l.swap $rD,$rA,$rB"
+     (+ OPC_CUST2 rD rA rB (f-resv-10-3 1) (f-resv-7-8 0))
+     (nop)
+     ()
+     )
 
 ; Misc instructions
 
@@ -1244,8 +1337,6 @@
     )
   )
 
-(cust-insn "1")
-(cust-insn "2")
 (cust-insn "3")
 (cust-insn "4")
 (cust-insn "5")
